pkg/reporter/github_reporter.go

```diff
@@ -0,0 +1,322 @@
+		}
+
+		f.FlagGitHubPullRequestNumber = prResponse.PullRequests[0].Number
+	}
+
+	logger.DebugContext(ctx, "computed pull request number", "computed_pull_request_number", f.FlagGitHubPullRequestNumber)
+
+	logURL, err := gc.ResolveJobLogsURL(ctx, f.FlagGitHubJob, f.FlagGitHubOwner, f.FlagGitHubRepo, f.FlagGitHubRunID)
+	if err != nil {
+		logger.WarnContext(ctx, "could not resolve direct url to job logs", "err", err)
+		logURL = fmt.Sprintf("%s/%s/%s/actions/runs/%d/attempts/%d", f.FlagGitHubServerURL, f.FlagGitHubOwner, f.FlagGitHubRepo, f.FlagGitHubRunID, f.FlagGitHubRunAttempt)
+	}
+
+	return &GitHubReporter{
+		gitHubClient: gc,
+		flags:        f,
+		logURL:       logURL,
+	}, nil
+}
+
+// validateFlags validates the required GitHub flags.
+func validateFlags(f flags.GitHubFlags) error {
+	var merr error
+
+	if f.FlagGitHubToken == "" && f.FlagGitHubAppID == "" {
+		merr = errors.Join(merr, fmt.Errorf("one of github token or github app id are required"))
+	}
+
+	if f.FlagGitHubToken != "" && f.FlagGitHubAppID != "" {
+		merr = errors.Join(merr, fmt.Errorf("only one of github token or github app id are allowed"))
+	}
+
+	if f.FlagGitHubAppID != "" && f.FlagGitHubAppInstallationID == "" {
+		merr = errors.Join(merr, fmt.Errorf("a github app installation id is required when using a github app id"))
+	}
+
+	if f.FlagGitHubAppID != "" && f.FlagGitHubAppPrivateKeyPEM == "" {
+		merr = errors.Join(merr, fmt.Errorf("a github app private key is required when using a github app id"))
+	}
+
+	if f.FlagGitHubOwner == "" {
+		merr = errors.Join(merr, fmt.Errorf("github owner is required"))
+	}
+
+	if f.FlagGitHubRepo == "" {
+		merr = errors.Join(merr, fmt.Errorf("github repo is required"))
+	}
+
+	if f.FlagGitHubPullRequestNumber <= 0 && f.FlagGitHubSHA == "" {
+		merr = errors.Join(merr, fmt.Errorf("one of github pull request number or github sha are required"))
+	}
+
+	if f.FlagGitHubServerURL == "" {
+		merr = errors.Join(merr, fmt.Errorf("github server url is required"))
+	}
+
+	if f.FlagGitHubRunID <= 0 {
+		merr = errors.Join(merr, fmt.Errorf("github run id is required"))
+	}
+
+	if f.FlagGitHubRunAttempt <= 0 {
+		merr = errors.Join(merr, fmt.Errorf("github run attempt is required"))
+	}
+
+	return merr
+}
+
+// CreateStatus implements the reporter Status function. It creates a new GitHub comment and persists the comment ID
+// in a private struct var for use by the UpdateStatus function.
+func (g *GitHubReporter) CreateStatus(ctx context.Context, p *Params) error {
+	msg, err := g.statusMessage(p)
+	if err != nil {
+		return fmt.Errorf("failed to generate status message")
+	}
+
+	startComment, err := g.gitHubClient.CreateIssueComment(
+		ctx,
+		g.flags.FlagGitHubOwner,
+		g.flags.FlagGitHubRepo,
+		g.flags.FlagGitHubPullRequestNumber,
+		msg.String(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to report: %w", err)
+	}
+
+	g.commentID = startComment.ID
```

Comment:
Should we be storing state on the reporter like this? Seems risky from an encapsulation perspective.

Reply:
For GitHub, we write a comment, then update it with the result, it gives immediate feedback when the run starts and ends. We could remove the update and just write a final comment instead which would remove the need for this state and extra function.

Reply:
Changed the implementation to only include CreateStatus, we wont write the first informational comment anymore (this is ok and probably better).

Reply:
Right, but the reporter should be largely stateless. Think about trying to run it in parallel - you wouldn't want to be storing information in the reporter itself.